### PR TITLE
'new' Event namespace validate failed

### PR DESCRIPTION
**Pull request by h4ghhh**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
In events.k8s.io/v1, if  involvedObject.namespace  is  "" , type Event's validation will fail.
When event.InvolvedObject.Namespace is "",  event.Namespace is set "default", but compared with "kube-system".

```
func (recorder *recorderImpl) makeEvent(refRegarding *v1.ObjectReference, refRelated *v1.ObjectReference, timestamp metav1.MicroTime, eventtype, reason, message string, reportingController string, reportingInstance string, action string) 
  *eventsv1.Event {
	t := metav1.Time{Time: recorder.clock.Now()}
	namespace := refRegarding.Namespace
	if namespace == "" {
		namespace = metav1.NamespaceDefault
	}
}
```

#### Which issue(s) this PR fixes:
Fixes # https://github.com/kubernetes/kubernetes/issues/100119

#### Does this PR introduce a user-facing change?
```release-note
kube-apiserver: events created via the `events.k8s.io` API group for cluster-scoped objects are now permitted in the default namespace as well for compatibility with events clients and the `v1` API
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/383-new-event-api-ga-graduation
```docs

```

### CODE REVIEW

File: pkg/apis/core/validation/events.go [hunk 1]
```diff
@@ -140,7 +140,7 @@ func legacyValidateEvent(event *core.Event) field.ErrorList {
 		}
 
 	} else {
-		if len(event.InvolvedObject.Namespace) == 0 && event.Namespace != metav1.NamespaceSystem {
+		if len(event.InvolvedObject.Namespace) == 0 && event.Namespace != metav1.NamespaceDefault {
```
Comment (liggitt) [inline pkg/apis/core/validation/events.go:143]: /hold

should we fix the event producer? Won't this break existing clients that are currently working?

Comment (h4ghhh) [reply pkg/apis/core/validation/events.go:143]: > /hold
> 
> should we fix the event producer? Won't this break existing clients that are currently working?

If existing clients use events.k8s.io/v1 rather than core/v1, has this problem long been exposed?
Codes is introduced to adapt 'new' event, rather than 'old'.
Refer to line 132, where is compared with  NamespaceDefault. So I think fixing validate func is better.

Comment (h4ghhh) [reply pkg/apis/core/validation/events.go:143]: No matter 'new' or 'old', the Event's default namespace is always 'default'. 
Refer to line 132, where compared with NamespaceDefault. 
These codes are introduced as new type Event check. If existing clients use events.k8s.io/v1 rather than core/v1, has this problem long been exposed?
So I think fixing validate func is better.

**Comment by h4ghhh** [comment]
refer to https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/383-new-event-api-ga-graduation
Namespace in which Event will live will be equal to
- Namespace of Regarding object, if it's namespaced,
- NamespaceSystem, if it's not.
 I decide to fix event client. Thx.

File: pkg/apis/core/validation/events.go (hunk 1 above)
Comment (h4ghhh) [reply pkg/apis/core/validation/events.go]: refer to https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/383-new-event-api-ga-graduation
Namespace in which Event will live will be equal to

Namespace of Regarding object, if it's namespaced,
NamespaceSystem, if it's not.
I decide to fix event client. Thx.

**Comment by h4ghhh** [comment]
I'm modifying kube-controller-manger to use new Event API, but this problem must be solved first. 
Please check if there is anything that not has been considered.
/cc @kubernetes/api-reviewers

**Comment by liggitt** [comment]
/assign @wojtek-t 

I think this change looks good, but we need tests exercising the following:
* creation of an events.k8s.io/v1 event for a cluster-scoped object that would have caught this issue
* a test ensuring `kubectl describe` on a built-in cluster-scoped object finds events recorded by both the old and new events client

**Comment by h4ghhh** [comment]
I do some test, results are as follow
event1

event2

event3

File: staging/src/k8s.io/client-go/tools/events/event_recorder.go [hunk 2]
```diff
@@ -67,7 +67,7 @@ func (recorder *recorderImpl) makeEvent(refRegarding *v1.ObjectReference, refRel
 	t := metav1.Time{Time: recorder.clock.Now()}
 	namespace := refRegarding.Namespace
 	if namespace == "" {
-		namespace = metav1.NamespaceDefault
+		namespace = metav1.NamespaceSystem
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/tools/events/event_recorder.go:70]: Actually - we're using NamespaceDefault in the old events library:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/record/event.go#L365-L368

The question is what we want do to here:
1) Change to NamespaceSystem here (as you do)
2) Change the validation logic to actually validate NamespaceDefault:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/events.go#L143

(1) might be misleading to users, as events that were previously landing in default namespace, will now be landing system namespace.
OTOH (2) seems more natural to me (I think that kube-system is more namespace for Node-related events). It may not be the case for other cluster-scoped resources though (e.g. PVs).

Neither of those options is perfect, but my personal preference would be towards (1) as we should prefer consistency (if someone is used to have events in default namespace for last 20 releases, we probably shouldn't change that now).

But I would like to hear @liggitt opinion from API perspective (can we change validation logic)?

Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/events/event_recorder.go:70]: I think `events.k8s.io/v1` events for cluster-scoped resources should be created in the same namespace that core `v1` events were.

We should not tighten API validation in a way that would break existing clients that are successfully writing events.

First, add a standalone integration test that does the following:

1. exercises the following test matrix for an event for a cluster-scoped object:

  * operations: `create`, `update`
  * API versions: `v1`, `events.k8s.io/v1`
  * namespaces: `default`, `kube-system`

2. uses the core v1 and events.k8s.io event recorders to record an event for a cluster-scoped object

Then, update the events.k8s.io client to behave like the core v1 client, creating events in the same namespace.

Comment (wojtek-t) [reply staging/src/k8s.io/client-go/tools/events/event_recorder.go:70]: Thanks Jordan.
I now realized that I wrote:

>  but my personal preference would be towards (1) as we should prefer consistency (if someone is used to have events in default namespace for last 20 releases, we probably shouldn't change that now).

which doesn't make sense - I meant option (2) [the argument about consistency remains - it was for option (2)].

So basically - we don't need any client changes - the client libraries are consistent with each other.

We need to update the validation logic, because currently we enforce "kube-system", whereas both the old core/v1.Event and events.k8s.io/v1/Event are writing events to "default" namespace.

Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/events/event_recorder.go:70]: agreed. we still need the tests to prove we're not breaking requests that work today, and that our clients actually work.

Comment (wojtek-t) [reply staging/src/k8s.io/client-go/tools/events/event_recorder.go:70]: Sure - we absolutely need tests. I just wanted to ensure that we will be testing the right thing - the current change isn't what we want.

**Comment by wojtek-t** [comment]
@h4ghhh - will you have time to work on this?

**Comment by h4ghhh** [comment]
> @h4ghhh - will you have time to work on this?

So, I understand that we decide to modify validation logic, but I don't know how to test.
Could you give me any suggestions? Thanks.

**Comment by wojtek-t** [comment]
> So, I understand that we decide to modify validation logic, but I don't know how to test.
Could you give me any suggestions? Thanks.

Here is the example of the simple test that actually also creates event:
https://github.com/kubernetes/kubernetes/blob/master/test/integration/client/client_test.go#L465

What you want additionally is to exercise a number of testcases somewhat like here:
https://github.com/kubernetes/kubernetes/blob/master/test/integration/apiserver/apiserver_test.go#L715

And your parameters would be:
- namespace
- API version
[and for each combination you call both create and update]

@h4ghhh - If you can start something and update the PR when you're stuck (even if it won't work compile) so that we can discuss more specifics.

File: test/integration/client/client_test.go [hunk 3]
```diff
@@ -460,6 +460,112 @@ func TestAPIVersions(t *testing.T) {
...
+
+	for _, test := range testcases {
+		switch test.clientVersion {
+		case "corev1":
+			event := mkCoreEvent(test.clientVersion, test.namespace)
+			_, err := client.CoreV1().Events(event.Namespace).Create(context.TODO(), event, metav1.CreateOptions{})
+			if err != nil && !test.hasError {
+				t.Errorf("%v, call Create failed, expect has error: %v, but got: %v", test.name, test.hasError, err)
+			}
+			_, err = client.CoreV1().Events(event.Namespace).Update(context.TODO(), event, metav1.UpdateOptions{})
```
Comment (wojtek-t) [inline test/integration/client/client_test.go:551]: update should actually do something (otherwise it won't even be committed)

File: test/integration/client/client_test.go [hunk 4]
```diff
@@ -460,6 +460,112 @@ func TestAPIVersions(t *testing.T) {
...
+			ReportingInstance:   "test-1",
+			Reason:              fmt.Sprintf("event %v test", name),
+			Action:              "Testing",
+		}
+	}
+
+	testcases := []struct {
+		name          string
+		namespace     string
+		clientVersion string
```
[the remark this answers is not in the document]
Comment (h4ghhh) [reply test/integration/client/client_test.go:515]: In each case, clientVersion is either corev1 or evnetsv1. Isn't it so?

**Comment by h4ghhh** [comment]
@wojtek-t Hi, the test is what we want to? By the way, I don't understand why gofmt and govet check failed

File: test/integration/client/client_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (h4ghhh) [reply test/integration/client/client_test.go:523]: Events created by corev1 or events are different, and should the response be different? 
Or do you want to say that namespace is the same?

**Comment by h4ghhh** [comment]
/cc @wojtek-t 
Hi, I have added a IT testcase. Could you review that for me? 
Thanks.

**Comment by thockin** [comment]
I don't really know enough about this area to comment, but it has come up in sig-net.  Is this the right fix?  Is it dead or can it make progress?

**Comment by wojtek-t** [comment]
/assign @liggitt 

Can you please review and approve for the API changes?

File: pkg/apis/core/validation/events.go (hunk 1 above)
Comment (liggitt) [reply pkg/apis/core/validation/events.go:143]: fixing the producer is the correct approach, but this PR is still trying to change API validation

I created an integration test that tries to create events for cluster-scoped objects against kube-system and default namespaces, via `v1 Events` and `events.k8s.io/v1 Events`

```diff
diff --git a/test/integration/events/events_test.go b/test/integration/events/events_test.go
index bb8830076ef..3f931bb78dc 100644
--- a/test/integration/events/events_test.go
+++ b/test/integration/events/events_test.go
@@ -18,10 +18,12 @@ package events
 
 import (
 	"context"
+	"strings"
 	"testing"
 	"time"
 
 	v1 "k8s.io/api/core/v1"
+	eventsv1 "k8s.io/api/events/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	clientset "k8s.io/client-go/kubernetes"
 
@@ -92,4 +94,84 @@ func TestEventCompatibility(t *testing.T) {
 	if err != nil {
 		t.Fatalf("unexpected err: %v", err)
 	}
+
+	coreevents := []*v1.Event{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "pass-core-kube-system-cluster-scoped",
+				Namespace: "kube-system",
+			},
+			EventTime:           metav1.NewMicroTime(time.Now()),
+			Type:                "Warning",
+			Action:              "myaction",
+			Reason:              "myreason",
+			InvolvedObject:      v1.ObjectReference{Name: "foo", Namespace: "", Kind: "Node", APIVersion: "v1"},
+			ReportingController: "mycontroller",
+			ReportingInstance:   "myinstance",
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "fail-core-default-cluster-scoped",
+				Namespace: "default",
+			},
+			EventTime:           metav1.NewMicroTime(time.Now()),
+			Type:                "Warning",
+			Action:              "myaction",
+			Reason:              "myreason",
+			InvolvedObject:      v1.ObjectReference{Name: "foo", Namespace: "", Kind: "Node", APIVersion: "v1"},
+			ReportingController: "mycontroller",
+			ReportingInstance:   "myinstance",
+		},
+	}
+	for _, e := range coreevents {
+		t.Run(e.Name, func(t *testing.T) {
+			_, err := client.CoreV1().Events(e.Namespace).Create(context.TODO(), e, metav1.CreateOptions{})
+			if err == nil && !strings.HasPrefix(e.Name, "pass-") {
+				t.Fatalf("unexpected pass")
+			}
+			if err != nil && !strings.HasPrefix(e.Name, "fail-") {
+				t.Fatalf("unexpected error: %v", err)
+			}
+		})
+	}
+
+	v1events := []*eventsv1.Event{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "pass-events-kube-system-cluster-scoped",
+				Namespace: "kube-system",
+			},
+			EventTime:           metav1.NewMicroTime(time.Now()),
+			Type:                "Warning",
+			Action:              "myaction",
+			Reason:              "myreason",
+			Regarding:           v1.ObjectReference{Name: "foo", Namespace: "", Kind: "Node", APIVersion: "v1"},
+			ReportingController: "mycontroller",
+			ReportingInstance:   "myinstance",
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "fail-events-default-cluster-scoped",
+				Namespace: "default",
+			},
+			EventTime:           metav1.NewMicroTime(time.Now()),
+			Type:                "Warning",
+			Action:              "myaction",
+			Reason:              "myreason",
+			Regarding:           v1.ObjectReference{Name: "foo", Namespace: "", Kind: "Node", APIVersion: "v1"},
+			ReportingController: "mycontroller",
+			ReportingInstance:   "myinstance",
+		},
+	}
+	for _, e := range v1events {
+		t.Run(e.Name, func(t *testing.T) {
+			_, err := client.EventsV1().Events(e.Namespace).Create(context.TODO(), e, metav1.CreateOptions{})
+			if err == nil && !strings.HasPrefix(e.Name, "pass-") {
+				t.Fatalf("unexpected pass")
+			}
+			if err != nil && !strings.HasPrefix(e.Name, "fail-") {
+				t.Fatalf("unexpected error: %v", err)
+			}
+		})
+	}
 }
```

Running that against 1.19, 1.20, 1.21, and 1.22 showed that events for cluster-scoped objects could only be created in the kube-system namespace in 1.19+

Comment (liggitt) [reply pkg/apis/core/validation/events.go:143]: Revert the changes to this file, add the integration test, and fix the namespace used when recording events for cluster-scoped objects in the event producer

File: pkg/apis/core/validation/events_test.go [hunk 5]
```diff
@@ -254,7 +254,7 @@ func TestValidateEventForNewV1beta1Events(t *testing.T) {
 			Event: &core.Event{
 				ObjectMeta: metav1.ObjectMeta{
 					Name:      "test",
-					Namespace: metav1.NamespaceSystem,
+					Namespace: metav1.NamespaceDefault,
```
Comment (liggitt) [inline pkg/apis/core/validation/events_test.go:257]: Revert the changes to this file

File: pkg/apis/core/validation/events.go (hunk 1 above)
Comment (liggitt) [reply pkg/apis/core/validation/events.go:143]: oh... that test is setting EventTime unnecessarily when creating corev1 events

This updated version of the test shows that you could create events for cluster-scoped objects in *either* the kube-system or default namespaces depending on whether you specific event time:

```diff
diff --git a/test/integration/events/events_test.go b/test/integration/events/events_test.go
index bb8830076ef..54f4001fa34 100644
--- a/test/integration/events/events_test.go
+++ b/test/integration/events/events_test.go
@@ -18,10 +18,12 @@ package events
 
 import (
 	"context"
+	"strings"
 	"testing"
 	"time"
 
 	v1 "k8s.io/api/core/v1"
+	eventsv1 "k8s.io/api/events/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	clientset "k8s.io/client-go/kubernetes"
 
@@ -92,4 +94,108 @@ func TestEventCompatibility(t *testing.T) {
 	if err != nil {
 		t.Fatalf("unexpected err: %v", err)
 	}
+
+	coreevents := []*v1.Event{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "fail-core-kube-system-cluster-scoped-no-time",
+				Namespace: "kube-system",
+			},
+			Type:                "Warning",
+			Action:              "myaction",
+			Reason:              "myreason",
+			InvolvedObject:      v1.ObjectReference{Name: "foo", Namespace: "", Kind: "Node", APIVersion: "v1"},
+			ReportingController: "mycontroller",
+			ReportingInstance:   "myinstance",
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "pass-core-kube-system-cluster-scoped-with-time",
+				Namespace: "kube-system",
+			},
+			EventTime:           metav1.NewMicroTime(time.Now()),
+			Type:                "Warning",
+			Action:              "myaction",
+			Reason:              "myreason",
+			InvolvedObject:      v1.ObjectReference{Name: "foo", Namespace: "", Kind: "Node", APIVersion: "v1"},
+			ReportingController: "mycontroller",
+			ReportingInstance:   "myinstance",
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "pass-core-default-cluster-scoped-no-time",
+				Namespace: "default",
+			},
+			Type:                "Warning",
+			Action:              "myaction",
+			Reason:              "myreason",
+			InvolvedObject:      v1.ObjectReference{Name: "foo", Namespace: "", Kind: "Node", APIVersion: "v1"},
+			ReportingController: "mycontroller",
+			ReportingInstance:   "myinstance",
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "fail-core-default-cluster-scoped-with-time",
+				Namespace: "default",
+			},
+			EventTime:           metav1.NewMicroTime(time.Now()),
+			Type:                "Warning",
+			Action:              "myaction",
+			Reason:              "myreason",
+			InvolvedObject:      v1.ObjectReference{Name: "foo", Namespace: "", Kind: "Node", APIVersion: "v1"},
+			ReportingController: "mycontroller",
+			ReportingInstance:   "myinstance",
+		},
+	}
+	for _, e := range coreevents {
+		t.Run(e.Name, func(t *testing.T) {
+			_, err := client.CoreV1().Events(e.Namespace).Create(context.TODO(), e, metav1.CreateOptions{})
+			if err == nil && !strings.HasPrefix(e.Name, "pass-") {
+				t.Errorf("unexpected pass")
+			}
+			if err != nil && !strings.HasPrefix(e.Name, "fail-") {
+				t.Fatalf("unexpected error: %v", err)
+			}
+		})
+	}
+
+	v1events := []*eventsv1.Event{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "pass-events-kube-system-cluster-scoped",
+				Namespace: "kube-system",
+			},
+			EventTime:           metav1.NewMicroTime(time.Now()),
+			Type:                "Warning",
+			Action:              "myaction",
+			Reason:              "myreason",
+			Regarding:           v1.ObjectReference{Name: "foo", Namespace: "", Kind: "Node", APIVersion: "v1"},
+			ReportingController: "mycontroller",
+			ReportingInstance:   "myinstance",
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "fail-events-default-cluster-scoped",
+				Namespace: "default",
+			},
+			EventTime:           metav1.NewMicroTime(time.Now()),
+			Type:                "Warning",
+			Action:              "myaction",
+			Reason:              "myreason",
+			Regarding:           v1.ObjectReference{Name: "foo", Namespace: "", Kind: "Node", APIVersion: "v1"},
+			ReportingController: "mycontroller",
+			ReportingInstance:   "myinstance",
+		},
+	}
+	for _, e := range v1events {
+		t.Run(e.Name, func(t *testing.T) {
+			_, err := client.EventsV1().Events(e.Namespace).Create(context.TODO(), e, metav1.CreateOptions{})
+			if err == nil && !strings.HasPrefix(e.Name, "pass-") {
+				t.Errorf("unexpected pass")
+			}
+			if err != nil && !strings.HasPrefix(e.Name, "fail-") {
+				t.Fatalf("unexpected error: %v", err)
+			}
+		})
+	}
 }
```

Comment (liggitt) [reply pkg/apis/core/validation/events.go:143]: And https://github.com/kubernetes/kubernetes/pull/88815 in 1.18 made both event reporters collapse onto the default namespace.

**Comment by liggitt** [comment]
I need to think about what the right solution here is... we can't make server changes that will break event API writers that are currently working, since you can create kube-system and default namespace events for cluster-scoped objects via the corev1 API, I think we have to relax API validation and allow either of those namespaces.

File: pkg/apis/core/validation/events_test.go (hunk 5 above)
Comment (h4ghhh) [reply pkg/apis/core/validation/events_test.go:257]: As you and @wojtek-t  discussed, we decided to change the validation function. Isn't it ?

Comment (liggitt) [reply pkg/apis/core/validation/events_test.go:257]: as noted in https://github.com/kubernetes/kubernetes/pull/100125#issuecomment-912823869, I think we have to *relax* API validation to allow either the `kube-system` or `default` namespaces for events for cluster-scoped objects.

That means this unit test should not change (since NamespaceSystem would still be allowed), but additional tests should be added to ensure NamespaceDefault is *also* allowed

Comment (liggitt) [reply pkg/apis/core/validation/events_test.go:257]: > As you and @wojtek-t discussed, we decided to change the validation function. Isn't it ?

yes, in the API validation, where we currently only allow NamespaceSystem, we should allow NamespaceSystem or NamespaceDefault

**Comment by h4ghhh** [comment]
Hi, @liggitt @wojtek-t
Now,  in new event validation logic, both NamespaceSystem and NamespaceDefault can pass the check.
By the way, I restore the existing test cases, and fix the integrate test bug about updating the eventTime.

**Comment by liggitt** [comment]
the new integration test is failing:

```
--- FAIL: TestEventValidation (6.46s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x3918174]

goroutine 26072 [running]:
testing.tRunner.func1.2({0x4031d80, 0x7b4a980})
	/usr/local/go/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1212 +0x218
panic({0x4031d80, 0x7b4a980})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
k8s.io/kubernetes/test/integration/client.TestEventValidation(0xc02a60a340)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/client/client_test.go:558 +0x6f4
```

**Comment by liggitt** [comment]
I would suggest an integration test more like https://github.com/kubernetes/kubernetes/pull/100125#discussion_r702169628 which fully exercises both the default and kube-system namespaces via both the old and new APIs

**Comment by h4ghhh** [comment]
> the new integration test is failing:
> 
> ```
> --- FAIL: TestEventValidation (6.46s)
> panic: runtime error: invalid memory address or nil pointer dereference [recovered]
> 	panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x3918174]
> 
> goroutine 26072 [running]:
> testing.tRunner.func1.2({0x4031d80, 0x7b4a980})
> 	/usr/local/go/src/testing/testing.go:1209 +0x24e
> testing.tRunner.func1()
> 	/usr/local/go/src/testing/testing.go:1212 +0x218
> panic({0x4031d80, 0x7b4a980})
> 	/usr/local/go/src/runtime/panic.go:1038 +0x215
> k8s.io/kubernetes/test/integration/client.TestEventValidation(0xc02a60a340)
> 	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/client/client_test.go:558 +0x6f4
> ```

I fixed it. Thanks.

**Comment by h4ghhh** [comment]
> I would suggest an integration test more like [#100125 (comment)](https://github.com/kubernetes/kubernetes/pull/100125#discussion_r702169628) which fully exercises both the default and kube-system namespaces via both the old and new APIs

I added test cases for compatibility.

**Comment by wojtek-t** [comment]
@h4ghhh - please add a release note also
Once merged, I think we should cherrypick it. @liggitt - FYI

**Comment by h4ghhh** [comment]
> @h4ghhh - please add a release note also
> Once merged, I think we should cherrypick it. @liggitt - FYI

Added.

**Comment by liggitt** [comment]
/hold

actually, can you squash this to a single commit before merge? it will make cherry-picks easier

**Comment by h4ghhh** [comment]
> /hold
> 
> actually, can you squash this to a single commit before merge? it will make cherry-picks easier

ok

**Comment by liggitt** [comment]
please open cherry-picks against 1.20-1.22

**Comment by h4ghhh** [comment]
> please open cherry-picks against 1.20-1.22

ok
